sample/weighted.go

```diff
@@ -0,0 +1,98 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted provides sampling without replacement from a collection of items with
+// weighted probabilites.
+type Weighted struct {
+	weights []float64
+	heap    []float64
+	src     *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. Elements of w are modified
+// during sampling, so a copy should be passed if the user needs to keep the values.
+// If src is nil, rand.Rand is used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
```

Comment:
I do have a preference toward my suggestion. 

The type and its methods do not exist to modify the values in w, it exists to provide random integers, and currently has a side effect of modifying w. It's more more cognitively simple for the user to think "NewWeighted gives me a sandbox for sampling without replacement" than it is to have to worry about the state of the slice. Furthermore, at present the shared state isn't consistent. If the weights are set with ReweightAll, the slice is free to modify.  At much as possible, we should be limiting "long-range" state.